### PR TITLE
Extend file upload component example to demonstrate how large files c…

### PR DESCRIPTION
**Pull request by paulgconlon**
…an be chunked into parts and sent via a MultiPart encoder. This can be useful for large design files.

### CODE REVIEW

File: components/files/file_upload/file_upload.py [hunk 1]
```diff
@@ -45,25 +52,66 @@ def to_json(self):
...
+            break
+        yield part 
+
+def upload_file_multipart(a_server_config, a_site_id, a_file_upload_bean, media_type, encoding_type, a_headers):
+    url = "{0}/file/v1/sites/{1}/upload".format(a_server_config.to_url(), a_site_id)
+    logging.debug("Upload file to {}".format(url))
+
+    file_path = a_file_upload_bean.file_location + os.path.sep + a_file_upload_bean.file_name
+    part_index = 0
+    sitelink_max_part_size = 10485760
```
Comment (topcon-lrietberg) [inline components/files/file_upload/file_upload.py:71]: Define a const MAX_PART_SIZE =  10485760?

File: components/files/file_upload/file_upload.py [hunk 2]
```diff
@@ -45,25 +52,66 @@ def to_json(self):
...
+    for part in part_generator:
+        file_size = os.stat(file_path).st_size
+        part_total_count = math.ceil(file_size / sitelink_max_part_size)
+
+        logging.debug("Preparing data part index {} of {} (size {})".format(part_index, part_total_count, len(part)))
+
+        form_header = {
+            "upload-uuid": str(a_file_upload_bean.upload_uuid),
+            "upload-file-name": a_file_upload_bean.file_name,
+            "upload-file-sha1": sha1.hexdigest(),
```
Comment (topcon-lrietberg) [inline components/files/file_upload/file_upload.py:87]: please move sha1.hexdigest() outside of the for loop.

File: components/files/file_upload/file_upload.py [hunk 3]
```diff
@@ -45,25 +52,66 @@ def to_json(self):
...
+# This function is used to chunk the data if its size exceeds the maximum part size accpeted by the 
+# Sitelink3D v2 file service (and other services that accept file such as designfile)
+def read_parts(a_file_ptr, a_part_size=10485760):
+    while True:
+        part = a_file_ptr.read(a_part_size)
+        if not part:
+            break
+        yield part 
+
+def upload_file_multipart(a_server_config, a_site_id, a_file_upload_bean, media_type, encoding_type, a_headers):
```
Comment (topcon-lrietberg) [inline components/files/file_upload/file_upload.py:65]: media_type and encoding_type are not used anymore

File: components/files/file_upload/file_upload.py [hunk 4]
```diff
@@ -27,6 +29,11 @@ def path_up_to_last(a_last, a_inclusive=True, a_path=os.path.dirname(os.path.rea
 # This is the mechanism by which file versioning is achieved.
 #
 # Because the FileUploadbean is the start of the upload_uuid chain, it is generated on creation rather than passsed in as a client parameter.
+#
+# Sitelink3D v2 will reject file uploads larger than 10485760 bytes. In these cases, the repsponse will be of the form 
+# 400:{"success":false,"error":"Upload size (41455634) exceeds maximum size (10485760) for a single part upload","preventRetry":false}
+# For this reason, multi-part uploads should be used for files in excess of this size. This is demonstrated in the below example.
+#
```
Comment (topcon-lrietberg) [inline components/files/file_upload/file_upload.py:36]: FileUploadBean.to_json seems is not used anymore, may be remove FileUploadBean, or refactor upload_file_multipart to use FileUploadBean instead?

File: components/files/file_upload/file_upload.py (hunk 1 above)
Comment (paulgconlon) [reply components/files/file_upload/file_upload.py:71]: Done.

File: components/files/file_upload/file_upload.py (hunk 2 above)
Comment (paulgconlon) [reply components/files/file_upload/file_upload.py:87]: Good find.

File: components/files/file_upload/file_upload.py (hunk 3 above)
Comment (paulgconlon) [reply components/files/file_upload/file_upload.py:65]: removed.

File: components/files/file_upload/file_upload.py (hunk 4 above)
Comment (paulgconlon) [reply components/files/file_upload/file_upload.py:36]: Refactored as the bean class is used more extensively throughout the examples and should also be leveraged here.

**Review by topcon-lrietberg — APPROVED** [review]

**Pull request merged**